compiler/rustc_expand/src/base.rs

```diff
@@ -543,20 +543,24 @@ impl DummyResult {
     ///
     /// Use this as a return value after hitting any errors and
     /// calling `span_err`.
-    pub fn any(span: Span) -> Box<dyn MacResult + 'static> {
-        Box::new(DummyResult { is_error: true, span })
+    pub fn any(span: Span, guar: ErrorGuaranteed) -> Box<dyn MacResult + 'static> {
+        Box::new(DummyResult { kind: Err(guar), span })
     }
 
     /// Same as `any`, but must be a valid fragment, not error.
     pub fn any_valid(span: Span) -> Box<dyn MacResult + 'static> {
-        Box::new(DummyResult { is_error: false, span })
+        Box::new(DummyResult { kind: Ok(()), span })
     }
 
     /// A plain dummy expression.
-    pub fn raw_expr(sp: Span, is_error: bool) -> P<ast::Expr> {
+    pub fn raw_expr(sp: Span, kind: Result<(), ErrorGuaranteed>) -> P<ast::Expr> {
```

Comment:
Would `Option<ErrorGuaranteed>` be better than `Result<(), ErrorGuaranteed>`?

Reply:
I hesitated between `Option` and `Result` in several places, and I've just realized now that it's actually just `bool` → `Option<ErrorGuaranteed>` and `T` → `Result<T, ErrorGuaranteed>` conversions, only that certain functions had an implicit `()` returns.

I'm busy for the next few days so I'll probably do this (and the next suggestions) on Monday.